### ISSUE TITLE
Make completions configurable

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#836]
At some point after #74 we should make the completions configurable - e.g. there are people who don't want quickmark completion for `:open`.

**Comment by Enzime** [comment]
Are there any ideas for a generalizable way to do this? @rcorre

**Comment by rcorre** [comment]
This is a pretty generic-sounding issue, but the only specific examples I've seen have to do with choosing the categories in `:open` completion. I've been in favor of using different commands to provide different completions. For those who don't want quickmarks in `open`, we could have `open-url` or `open-history`. For buffer/url, we could have `open-or-switch` (or some more clever name). This also helps disambiguate arguments -- `open-or-switch` could have special logic to interpret buffer ids, `open-history` would _not_ try to interpret `foobar` as a quickmark.

Two other options, which I don't prefer, are:

1. Use flags to control completion (e.g. `:open --no-quickmarks`). Flags are conventionally arguments to the command. I don't like the idea of no-op flags arguments that must be ignored in the function and exist solely for completion.

2. Have a setting `completion.command.open.categories` = ['history', 'bookmark', 'buffer']`. Huh. Actually I might not hate this one. Given a recent discussion on the mailing list I could also imagine `completion.command.open.mode = 'menu' # or 'longest'. My concern is that this is overkill and unintuitive. I'm biased because I've worked on completion a lot, but I wonder if anyone not familiar with the implementation would even understand such a setting.

**Comment by The-Compiler** [comment]
I'd rather have not many similar commands, it's the reason qutebrowser has `:open -w` and `-t` instead of stuff like `:winopen` and `:tabopen` :wink:

The nice thing about having flags is that if you e.g. have `:open --no-quickmarks`, `:open` can actually ignore quickmarks, and e.g. doing `:open git` with a quickmark `git` will actually open a search - I think there's actually an issue about this somewhere.

The `categories` setting would definitely be the most flexible, but the command always also has to handle those cases.

**Comment by Enzime** [comment]
Not too sure how to implement `--no-quickmarks` with regards to the completions, because they're initialized without knowing what the command is.

**Comment by The-Compiler** [comment]
@Enzime Yeah, you'd first need to find a way to pass the command (or partially parsed commandline) to completion functions first.

**Comment by rcorre** [comment]
> @Enzime Yeah, you'd first need to find a way to pass the command (or partially parsed commandline) to completion functions first.

Stick it in `CompletionInfo`. It could also be used to filter the command suggestion for `bind --mode MODE`.

**Comment by rcorre** [comment]
Ah, I remember the challenge here. We use [`argparser.parse_args`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args), which fails if you don't have a semantically correct command. For example, `bind --mode command` will not give you `{"mode": "command"}`, but rather will fail because you have not provided the required `key` parameter. We might need to roll our own parser :grimacing:

**Comment by Enzime** [comment]
The completion function only runs once to initialize the completer model, so I modified `qutebrowser/completion/completer.py::_update_completion` to create a new completer model every time the completions are updated and pass the command in. However, I'm not sure if this will become a performance concern in the future.

One optimization I can think of is having a secondary function to determine if the model should be reinitialized (this could be combined with incremental parsing to switch argument completions).

**Comment by rcorre** [comment]
> The completion function only runs once to initialize the completer model

ah, yes. As the code stands right now, changing the flags wouldn't generate a new set of completions.

> However, I'm not sure if this will become a performance concern in the future

The history model runs a sql query every time the pattern changes. However, it currently changes the bound parameter and re-runs the same prepared query rather than instantiating a new query. See https://stackoverflow.com/a/1703234/1435461. We could store that prepared query object somewhere more persistent than the `HistoryCategory` if needed ...
